src/Features/Core/Portable/ExtractClass/AbstractExtractClassRefactoringProvider.cs

```diff
@@ -0,0 +1,107 @@
+// See the LICENSE file in the project root for more information.
+
+#nullable enable 
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.PullMemberUp;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.ExtractClass
+{
+    internal abstract class AbstractExtractClassRefactoringProvider : CodeRefactoringProvider
+    {
+        public AbstractExtractClassRefactoringProvider()
+        {
+        }
+
+        public AbstractExtractClassRefactoringProvider(IExtractClassOptionsService service)
+        {
+            _optionsService = service;
+        }
+
+        private readonly IExtractClassOptionsService? _optionsService;
+
+        protected abstract Task<SyntaxNode?> GetSelectedNodeAsync(CodeRefactoringContext context);
+        protected abstract Task<SyntaxNode?> GetSelectedClassDeclarationAsync(CodeRefactoringContext context);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var optionsService = _optionsService ?? context.Document.Project.Solution.Workspace.Services.GetService<IExtractClassOptionsService>();
+            if (optionsService is null)
+            {
+                return;
+            }
+
+            // If we register the action on a class node, no need to find selected members. Just allow
+            // the action to be invoked with the dialog 
+            if (await TryRegisterClassActionAsync(context, optionsService).ConfigureAwait(false))
+            {
+                return;
+            }
+
+            await RegisterMemberActionAsync(context, optionsService).ConfigureAwait(false);
+        }
+
+        private async Task RegisterMemberActionAsync(CodeRefactoringContext context, IExtractClassOptionsService optionsService)
+        {
+            var selectedMemberNode = await GetSelectedNodeAsync(context).ConfigureAwait(false);
+            if (selectedMemberNode is null)
+            {
+                return;
+            }
+
+            var (document, span, cancellationToken) = context;
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var selectedMember = semanticModel.GetDeclaredSymbol(selectedMemberNode);
+            if (selectedMember is null || selectedMember.ContainingType is null)
+            {
+                return;
+            }
+
+            // Use same logic as pull members up for determining if a selected member
+            // is valid to be moved into a base
+            if (!MemberAndDestinationValidator.IsMemberValid(selectedMember))
+            {
+                return;
+            }
+
+            var containingType = selectedMember.ContainingType;
+
+            // Can't extract to a new type if there's already a base. Maybe
+            // in the future we could inject a new type inbetween base and
+            // current
```

Comment:
yes, it seems like that woudl be nice to have and expected.  but def doesn't have to be there initially.

Reply:
My plan is just to wait and see if people ask for this feature moving forward, and we can improve after that.